File: src/Conduit.Rag/Parsing/Languages/GoParser.cs
using System.Text.RegularExpressions;

namespace Conduit.Rag.Parsing.Languages;

/// <summary>
/// Regex-based parser for Go source files.
/// Detects func declarations, struct types, and interface types.
/// Exported = uppercase first letter.
/// </summary>
public sealed partial class GoParser : ICodeParser
{
    public bool CanParse(string extension)
        => extension.Equals(".go", StringComparison.OrdinalIgnoreCase);

    public IReadOnlyList<CodeUnit> Parse(string content, string filePath)
    {
        try
        {
            return ParseInternal(content, filePath);
        }
        catch
        {
            return [];
        }
    }

    private static IReadOnlyList<CodeUnit> ParseInternal(string content, string filePath)
    {
        var units   = new List<CodeUnit>();
        var lines   = content.Split('\n');
        var package = ExtractPackage(lines);

        for (var i = 0; i < lines.Length; i++)
        {
            var line = lines[i];

            // func (receiver) Name(...) ...
            var funcMatch = FuncRegex().Match(line);
            if (funcMatch.Success)
            {
                var name            = funcMatch.Groups["name"].Value;
                var receiver        = funcMatch.Groups["receiver"].Value.Trim();
                var declarationLine = i;                                    // capture before i is moved
                var block           = ParserUtils.ExtractBraceBlock(lines, i, out var endLine);
                i = endLine;

                units.Add(new CodeUnit
                {
                    Kind          = CodeUnitKind.Function,
                    Name          = name,
                    ContainerName = string.IsNullOrEmpty(receiver) ? null : StripPointer(receiver),
                    Namespace     = package,
                    Signature     = line.Trim(),
                    IsPublic      = name.Length > 0 && char.IsUpper(name[0]),
                    DocComment    = ExtractGoDoc(lines, declarationLine),   // use declaration line
                    FullText      = block,
                    Language      = "Go",
                    FilePath      = filePath,
                });
                continue;
            }

            // type Name struct { ... }
            var structMatch = StructRegex().Match(line);
            if (structMatch.Success)
            {
                var name            = structMatch.Groups["name"].Value;
                var declarationLine = i;
                var block           = ParserUtils.ExtractBraceBlock(lines, i, out var endLine);
                i = endLine;

                units.Add(new CodeUnit
                {
                    Kind          = CodeUnitKind.Struct,
                    Name          = name,
                    ContainerName = null,
                    Namespace     = package,
                    Signature     = null,
                    IsPublic      = name.Length > 0 && char.IsUpper(name[0]),
                    DocComment    = ExtractGoDoc(lines, declarationLine),
                    FullText      = block,
                    Language      = "Go",
                    FilePath      = filePath,
                });
                continue;
            }

            // type Name interface { ... }
            var ifaceMatch = InterfaceRegex().Match(line);
            if (ifaceMatch.Success)
            {
                var name            = ifaceMatch.Groups["name"].Value;
                var declarationLine = i;
                var block           = ParserUtils.ExtractBraceBlock(lines, i, out var endLine);
                i = endLine;

                units.Add(new CodeUnit
                {
                    Kind          = CodeUnitKind.Interface,
                    Name          = name,
                    ContainerName = null,
                    Namespace     = package,
                    Signature     = null,
                    IsPublic      = name.Length > 0 && char.IsUpper(name[0]),
                    DocComment    = ExtractGoDoc(lines, declarationLine),
                    FullText      = block,
                    Language      = "Go",
                    FilePath      = filePath,
                });
            }
        }

        return units;
    }

    /// <summary>Scans upward from <paramref name="declarationLine"/> collecting // comment lines.</summary>
    private static string? ExtractGoDoc(string[] lines, int declarationLine)
    {
        var docLines = new List<string>();
        var i        = declarationLine - 1;

        while (i >= 0 && lines[i].TrimStart().StartsWith("//"))
        {
            docLines.Insert(0, lines[i].TrimStart().TrimStart('/').Trim());
            i--;
        }

        return docLines.Count > 0 ? string.Join("\n", docLines) : null;
    }

    private static string? ExtractPackage(string[] lines)
    {
        foreach (var line in lines)
        {
            var m = PackageRegex().Match(line);
            if (m.Success) return m.Groups["pkg"].Value;
        }
        return null;
    }

    private static string StripPointer(string receiver)
    {
        // e.g. "r *MyType" → "MyType"
        return receiver.Split(' ').Last().TrimStart('*');
    }

    [GeneratedRegex(@"^func\s+(\((?<receiver>[^)]*)\)\s+)?(?<name>[A-Za-z_]\w*)\s*\(", RegexOptions.Compiled)]
    private static partial Regex FuncRegex();

    [GeneratedRegex(@"^type\s+(?<name>[A-Za-z_]\w*)\s+struct\b", RegexOptions.Compiled)]
    private static partial Regex StructRegex();

    [GeneratedRegex(@"^type\s+(?<name>[A-Za-z_]\w*)\s+interface\b", RegexOptions.Compiled)]
    private static partial Regex InterfaceRegex();

    [GeneratedRegex(@"^package\s+(?<pkg>\w+)", RegexOptions.Compiled)]
    private static partial Regex PackageRegex();
}

File: app/parsing/go_parser.py
from __future__ import annotations

import re
from app.models import CodeUnit, CodeUnitKind
from app.parsing.base import CodeParser

_PACKAGE_RE = re.compile(r"^package\s+(\w+)", re.MULTILINE)
_FUNC_RE = re.compile(
    r"^func\s+(?:\((?P<receiver>[^)]+)\)\s+)?(?P<name>\w+)\s*\((?P<params>[^)]*)\)",
    re.MULTILINE,
)
_STRUCT_RE = re.compile(r"^type\s+(?P<name>\w+)\s+struct\b", re.MULTILINE)
_INTERFACE_RE = re.compile(r"^type\s+(?P<name>\w+)\s+interface\b", re.MULTILINE)
_COMMENT_BLOCK_RE = re.compile(r"((?:\s*//[^\n]*\n)+)$")


def _preceding_comment(content: str, match_start: int) -> str:
    before = content[:match_start]
    m = _COMMENT_BLOCK_RE.search(before)
    if m:
        lines = [re.sub(r"^\s*//\s?", "", l) for l in m.group(1).splitlines()]
        return " ".join(l for l in lines if l.strip())
    return ""


def _extract_block(content: str, start: int, max_lines: int = 80) -> str:
    return "\n".join(content[start:].splitlines()[:max_lines])


class GoParser(CodeParser):
    def can_parse(self, extension: str) -> bool:
        return extension.lower() == ".go"

    def parse(self, content: str, file_path: str) -> list[CodeUnit]:
        try:
            return self._parse(content, file_path)
        except Exception:
            return []

    def _parse(self, content: str, file_path: str) -> list[CodeUnit]:
        units: list[CodeUnit] = []

        pkg_match = _PACKAGE_RE.search(content)
        namespace = pkg_match.group(1) if pkg_match else None

        for m in _FUNC_RE.finditer(content):
            name = m.group("name")
            receiver_raw = m.group("receiver") or ""
            receiver = re.sub(r"\*", "", receiver_raw.split()[-1]).strip() if receiver_raw else None
            doc = _preceding_comment(content, m.start())
            params = m.group("params")
            block = _extract_block(content, m.start())

            units.append(CodeUnit(
                kind=CodeUnitKind.FUNCTION,
                name=name,
                container_name=receiver,
                namespace=namespace,
                signature=f"func {name}({params})",
                is_public=name[0].isupper() if name else False,
                doc_comment=doc or None,
                full_text=block[:3000],
                language="Go",
                file_path=file_path,
            ))

        for pattern, kind in [(_STRUCT_RE, CodeUnitKind.STRUCT), (_INTERFACE_RE, CodeUnitKind.INTERFACE)]:
            for m in pattern.finditer(content):
                name = m.group("name")
                doc = _preceding_comment(content, m.start())
                block = _extract_block(content, m.start())
                units.append(CodeUnit(
                    kind=kind,
                    name=name,
                    namespace=namespace,
                    is_public=name[0].isupper() if name else False,
                    doc_comment=doc or None,
                    full_text=block[:3000],
                    language="Go",
                    file_path=file_path,
                ))

        if not units:
            units.append(CodeUnit(
                kind=CodeUnitKind.FILE,
                name=file_path.split("/")[-1].split("\\")[-1],
                namespace=namespace,
                full_text=content[:4000],
                language="Go",
                file_path=file_path,
            ))

        return units
